rust/user-info-fetcher/src/backend/keycloak.rs

```diff
@@ -0,0 +1,180 @@
+///
+/// [users]: https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html#_get_adminrealmsrealmusers
+/// [user-by-id]: https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html#_get_adminrealmsrealmusersid
+/// [UserRepresentation]: https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html#UserRepresentation
+/// [issue-20292]: https://github.com/keycloak/keycloak/issues/20294
+#[derive(Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct UserMetadata {
+    id: String,
+    username: String,
+    #[serde(default)]
+    attributes: HashMap<String, Vec<String>>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GroupMembership {
+    path: String,
+}
+
+pub(crate) async fn get_user_info(
+    req: &UserInfoRequest,
+    http: &reqwest::Client,
+    credentials: &Credentials,
+    config: &crd::KeycloakBackend,
+) -> Result<UserInfo, Error> {
+    let crd::KeycloakBackend {
+        client_credentials_secret: _,
+        admin_realm,
+        user_realm,
+        hostname,
+        port,
+        root_path,
+        tls,
+    } = config;
+
+    // We re-use existent functionality from operator-rs, besides it being a bit of miss-use.
+    // Some attributes (such as principal_claim) are irrelevant, and will not be read by the code-flow we trigger.
+    let wrapping_auth_provider = oidc::AuthenticationProvider::new(
+        hostname.clone(),
+        *port,
+        root_path.clone(),
+        tls.clone(),
+        String::new(),
+        Vec::new(),
+        None,
+    );
+    let keycloak_url = wrapping_auth_provider
+        .endpoint_url()
+        .context(ParseOidcEndpointUrlSnafu)?;
+
+    let authn = send_json_request::<OAuthResponse>(
+        http.post(
+            keycloak_url
+                .join(&format!(
+                    "realms/{admin_realm}/protocol/openid-connect/token"
+                ))
+                .context(ConstructOidcEndpointPathSnafu)?,
+        )
+        .basic_auth(&credentials.username, Some(&credentials.password))
+        .form(&[("grant_type", "client_credentials")]),
+    )
+    .await
+    .context(LogInSnafu)?;
+
+    let users_base_url = keycloak_url
+        .join(&format!("admin/realms/{user_realm}/users/"))
+        .context(ConstructOidcEndpointPathSnafu)?;
+
+    let user_info = match req {
+        UserInfoRequest::UserInfoRequestById(req) => {
+            let user_id = req.id.clone();
+            send_json_request::<UserMetadata>(
+                http.get(
+                    users_base_url
+                        .join(&req.id)
+                        .context(ConstructOidcEndpointPathSnafu)?,
+                )
+                .bearer_auth(&authn.access_token),
+            )
+            .await
+            .context(UserNotFoundByIdSnafu { user_id })?
+        }
+        UserInfoRequest::UserInfoRequestByName(req) => {
+            let user_name = &req.username;
```

Comment:
```suggestion
            let username = &req.username;
```

Username != User's (display? legal?) name.

Reply:
Fixed in 88c0e35